### DISCUSSION TITLE
Drag and drop + subscriptions

### DISCUSSION THREAD

**Discussion by bubbaspaarx** [urql-graphql/urql#2715] [Q&A]
Hi there, 

I'm not entirely sure what the solution is to this one and it's been driving me mildly insane the last month. 

We have built a multi lane drag and drop feature using react beautiful dnd and it was all working perfectly until we added subscriptions in. Premise is that the board will be used by multiple people so therefore needs to be live. 

The main issue is that there are 2 components to our boards, the first being that the card in the lane has a status and the board itself has an order. For us to maintain the order and the status of each card, we have to make 2 mutations when a card is moved. This causes a headache and race condition between the 2 request. 
If the order comes back before the cards status then the card may flash position temporarily before the new order is available and it then corrects itself. 
We have utilised optimistic rendering but this just makes the issue faster.

Is there a way to prevent the return value of a mutation conditionally. i.e. don't return anything until x happens or a way to combine mutations?

We also sometimes only need to make 1 mutation.  For instance, if the order doesn't change but the status of the card does then only an status mutation is required, the same as if we only change the order within a lane then only an order mutation is required. 

Lots of moving parts to consider and wonder if there's a better approach. 

Thank you in advance.

**Comment by frederikhors** [comment]
Can you post an example? A reproduction?

**Reply by bubbaspaarx** [reply]
HI @frederikhors. I can show you the front end aspect but imagine the backend has a lot to play in this scenario too. 

urlq query and subscription
```
  const [listResult] = useQuery({
    query: listQuery,
    variables: { id: selectedList?.id },
    requestPolicy: 'cache-first'
  })

  const [_listSubscription] = useSubscription({
    query: listSubscriptionQuery,
    variables: { listId: selectedList?.id },
    requestPolicy: 'cache-first'
  })
```
 
 Example drag and drop function with mutations
```
const onDragEnd = (result, boardLanes) => {

 >>>>>D&D logic<<<<<<

if (source.droppableId !== destination.droppableId) {

>>>>>More list ordering logic<<<<<

// mutation for updating the order
updateOrder({ listId: currentList?.id, order: newOrder })

// mutation for updating the individual card
updateListCardl({
   listCardId: listCard.id,
   status: destinationLane.name
 })

} else {
>>>>>More list ordering logic<<<<<

// mutation for updating the order
updateOrder({ listId: currentList?.id, order: newOrder })

}
```

Urql optimistic cache update
```
 optimistic: {
    updateOrder: (variables, cache, _info) => {
      return {
        __typename: 'List',
        id: variables.listId,
        order: variables.order
      }
    },
    updateCard: (variables, cache, info) => {
      return {
        __typename: 'Card',
        id: variables.cardId,
        status: variables.status
      }
  }
}
```

Let me know if this makes any sense. Stripped out a lot of the basic ordering functionality as it's not particularly relevant to the whole situation. 
And for clarity, ordering across the lanes is working but the positions flicker (inconsistently) and can cause flickers on both the side the user is moving the card and the side receiving the subscription. 

In the backend we are calling the subscription to update the list both when we update the card and update the list as the other users need to see changes for both.

**Comment by JoviDeCroock** [comment]
This is actually a peculiar scenario to me given that we have invested a lot of time in ensuring updates are [commutative](https://formidable.com/open-source/urql/docs/graphcache/normalized-caching/#deterministic-cache-updates).

One question I would have is are subscriptions broadcasted to their initiator as well or only to the other people, i.e. are you ignoring subscription triggers for the person who made the update?

The reason behind the above questions is that subscriptions can't really be applied commutatively. If you aren't broadcasting the subscription triggers to the initiator then this should not be happening, one thing to try is to remove the other attributes from the mutation, i.e. only ask the property you have altered so that `updateOrder` can't return a stale `status` and `updateStatus` can't return a stale `order`.

Is there a reason behind having two distinct mutations for an update to one type? I would assume that updating them in one mutation wouldn't cause any issue as the subscription seems to be listening to general updates for the Card either way.

**Reply by bubbaspaarx** [reply]
I'm really sorry for not responding to this earlier. I must have completely missed it. 

We aren't ignoring the subscriptions from our own triggers unfortunately. Is this something we can do with Urql or are you suggesting a tweak to our api for this?

We actually did go down the route of updating this through a singular mutation but it didn't fully resolve the situation. I definitely don't think the drag and drop package we are using helps the situation as Urql otherwise works flawlessly throughout the system

**Reply by JoviDeCroock** [reply]
I'd say you can change that in the back-end by not emitting for your own updates 😅 it's a bit hard with these race conditions I do agree but generally you are receiving a double update one form the subscription and one from the mutation so the behaviour itself makes sense.
